### PR TITLE
dfu: allow flash_img to be used without mcuboot.c

### PR DESCRIPTION
**Pull request by hakonfam**
This module is useful without mcuboot, don't tie it to mcuboot.

Please see this https://github.com/zephyrproject-rtos/zephyr/pull/23290 before reviewing this PR.

### CODE REVIEW

File: subsys/dfu/Kconfig [hunk 1]
```diff
@@ -14,20 +14,34 @@ menuconfig IMG_MANAGER
 
 if IMG_MANAGER
 
-choice
-	prompt "Image manager"
-	default MCUBOOT_IMG_MANAGER
+config FLASH_IMG
+	bool "Flash Image API"
```
Comment (nvlsianpu) [inline subsys/dfu/Kconfig]: need some better prompt here which could  help with understanding what this module really does: Flash image collection API?

File: subsys/dfu/Kconfig [hunk 2]
```diff
@@ -14,20 +14,34 @@ menuconfig IMG_MANAGER
...
+	help
+	  If enabled, flash is erased as necessary when receiving new firmware,
+	  instead of erasing the whole image slot at once. This is necessary
+	  on some hardware that has long erase times, to prevent long wait
+	  times at the beginning of the DFU process.
 
 config MCUBOOT_IMG_MANAGER
 	bool "Image manager for mcuboot"
-	select FLASH_MAP
+	select FLASH_IMG
```
Comment (nvlsianpu) [inline subsys/dfu/Kconfig]: Is this for backward compatibility? -I believe this not really required for meet dependency between module. Maybe better to make `FLASH_IMG=y` by default depend on this property?

File: subsys/dfu/img_util/flash_img.c [hunk 3]
```diff
@@ -234,6 +234,8 @@ int flash_img_init(struct flash_img_context *ctx)
 #ifdef CONFIG_IMG_ERASE_PROGRESSIVELY
 	ctx->off_last = -1;
 #endif
+#ifdef FLASH_AREA_IMAGE_SECONDARY
```
Comment (nvlsianpu) [inline subsys/dfu/img_util/flash_img.c]: why this change is here?

Comment (hakonfam) [reply subsys/dfu/img_util/flash_img.c]: I will remove this, it sort of depends on https://github.com/zephyrproject-rtos/zephyr/pull/23218 to make sense. 

In essence it allows the file to compile without having the mcuboot defines present.

File: subsys/dfu/Kconfig (hunk 1 above)
Comment (hakonfam) [reply subsys/dfu/Kconfig]: I agree. 

Ideally the flash_img.c file should be moved out of the dfu subsystem altogether, and renamed to something more generic. Like `flash_bufferd_write.c`, as this is actually what it does, it is not related to DFU at all.

I am willing to make this change if it is deemed valid by you guys.

Comment (nvlsianpu) [reply subsys/dfu/Kconfig]: For me sounds OK.
I can suggest subsys/storage/flash_buffred_writer[or fbw] path.

Comment (mbolivar) [reply subsys/dfu/Kconfig]: SGTM though the "buffered" typo should be fixed.

Comment (hakonfam) [reply subsys/dfu/Kconfig]: https://github.com/zephyrproject-rtos/zephyr/pull/23290

Here is a draft pull request with the generic module. I see that this is definitely something we need (think download to external flash, loading from external flash to internal flash, loading image from application core flash to network core flash on 53).

Comment (mbolivar) [reply subsys/dfu/Kconfig]: > #23290
> 
> Here is a draft pull request with the generic module. I see that this is definitely something we need (think download to external flash, loading from external flash to internal flash, loading image from application core flash to network core flash on 53).

Thanks a lot! I've left some comments there. I think this is a great direction to move in.

I hope you're not too upset with me for widening the scope to include revisiting the flash_map.h API.

**Comment by hakonfam** [comment]
Close this as https://github.com/zephyrproject-rtos/zephyr/pull/23290 will be used instead
